recipe-server/client/actions/show-heartbeat/index.js

```diff
@@ -16,35 +19,218 @@ export default class ShowHeartbeatAction extends Action {
+    return typeof lastShown !== 'undefined' ?
+      parseFloat(lastShown) : null;
+  }
+
+  /**
    * Checks when this survey was last shown,
    * and returns a boolean indicating if the
-   * user has seen this survey already or not.
+   * user has ever seen this survey or not.
+   *
+   * @async
+   * @return {Boolean}  Has the survey ever been shown?
    */
-  async surveyHasShown() {
-    const lastShown = await this.storage.getItem('lastShown');
+  async hasShownBefore() {
+    const lastShown = await this.getLastShown();
     // If no survey has been shown, lastShown will be falsey.
     return !!lastShown;
   }
 
+  /**
+   * Determines if this heartbeat was shown
+   * at least x days ago.
+   *
+   * @param  {Number}  days Days ago to check
+   * @return {boolean}      Has prompt been shown by that date?
+   */
+  async shownAtleastDaysAgo(days) {
+    const hasShown = await this.hasShownBefore();
+
+    if (!hasShown) {
+      return false;
+    }
+
+    // get timestamp of last shown
+    const timeLastShown = await this.getLastShown();
+
+    // get the difference between now and then
+    const timeElapsed = Date.now() - timeLastShown;
+
+    // time limit is the number of days passed in
+    // converted into milliseconds
+    const timeLimit = ONE_DAY * days;
+
+    // if the diff is smaller than the limit,
+    // that means that the last time the user saw the prompt
+    // was less than the `days` passed in
+    return timeElapsed < timeLimit;
+  }
+
+  /**
+   * Simple function to read the lastInteraction
+   * timestamp (if any) from local storage.
+   * @return {number}   Timestamp of last prompt interaction (if any)
+   */
+  async getLastInteraction() {
+    const lastInteraction = await this.storage.getItem('lastInteraction');
+
+    return typeof lastInteraction !== 'undefined' ?
+      parseFloat(lastInteraction) : null;
+  }
+
+  /**
+   * Gets the timestamp of the last prompt interaction,
+   * and returns the time (in ms) since then.
+   *
+   * @async
+   * @return {number}
+   */
+  async sinceLastInteraction() {
+    const lastInteraction = await this.getLastInteraction();
+
+    return typeof lastInteraction !== 'undefined' ?
+      Date.now() - lastInteraction : null;
+  }
+
+  /**
+   * Checks when the survey prompt last had
+   * interaction from the user (if ever),
+   * and returns a boolean indicating if the
+   * user has ever had interaction
+   *
+   * @async
+   * @return {Boolean}  Has the survey ever had interaction?
+   */
+  async hasHadInteraction() {
+    const lastInteraction = await this.getLastInteraction();
+    return !!lastInteraction;
+  }
+
+  /**
+   * Checks the repeat argument for this recipe,
+   * then determines if the recipe can be qualified as 'ran'.
+   * This ultimately decides if the prompt is shown at all
+   * to the end user.
+   *
+   * @return {boolean}        Has the heartbeat been shown?
+   */
+  async heartbeatHasRan() {
```

Comment:
I'm confused. This function checks the recipe to see if it's fulfilled the requirements to be considered 'shown' (that `once` has been shown once, that `nag` has been interacted with, etc). It doesn't determine if anything _should_ run - it just checks if the heartbeat has met its specific criteria for being 'ran'.

The function `shouldNotExecute` (Line 214) determines if anything should run (or specifically, shouldn't). It checks normandy's testing state, the global heartbeat timer, and then checks if this specific heartbeat has ran (via `heartbeatHasRan`).

Perhaps the naming convention needs to be changed for some of these functions?